### PR TITLE
 Layout.FromMethod that supports typed Layout #5532 

### LINKED ISSUE #5532: Layout.FromMethod that supports typed Layout
Implement helper to return type `Layout<T>`

```c#
public static Layout<T> FromMethod<T>(Func<LogEventInfo, T> layoutMethod, LayoutRenderOptions options = LayoutRenderOptions.None)
{
   ...
}
```

### PR DESCRIPTION
**Pull request by smnsht**
Trying to resolve #5532

### CODE REVIEW

File: tests/NLog.UnitTests/Layouts/LayoutTypedTests.cs [hunk 1]
```diff
@@ -42,11 +44,11 @@ public class LayoutTypedTests : NLogTestBase
         [Fact]
         public void LayoutFixedIntValueTest()
         {
-            // Arrange
+            // Arrange
```
Comment (snakefoot) [inline tests/NLog.UnitTests/Layouts/LayoutTypedTests.cs:47]: Possible that you can remove these extra white-spaces?

**Comment by snakefoot** [comment]
Can you also copy the `Url`-value-tests, and verify they will work with `Layout.FromMethod` ?

File: tests/NLog.UnitTests/Layouts/LayoutTypedTests.cs (hunk 1 above)
Comment (smnsht) [reply tests/NLog.UnitTests/Layouts/LayoutTypedTests.cs]: Done.

I've noticed file headers retain trailing spaces, so, I turn off an extension (trailing white spaces) in Visual Studio when working on NLog.

What would be the correct way to handle the situation?

**Comment by smnsht** [comment]
> Can you also copy the `Url`-value-tests, and verify they will work with `Layout.FromMethod` ?

Done.

`LayoutFixedEmptyUrlTest()` should be skipped, because It is not relevant for typed `FromMethod()`

Is that correct?

**Comment by smnsht** [comment]
For more involved changes I will be available on the weekend (Fri-Sat)

**Comment by snakefoot** [comment]
Something similar to ``LayoutFixedEmptyUrlTest` could be relevant to verify what happens if the Func-method returns `null`

File: tests/NLog.UnitTests/Layouts/LayoutTypedTests.cs (hunk 1 above)
Comment (snakefoot) [reply tests/NLog.UnitTests/Layouts/LayoutTypedTests.cs]: Yes I have notice that other developers has extensions included that does different bonus formatting on save-file.

Guess one could "optimize" the NLog-repo to be more friendly towards bonus-formatting-developers.

**Comment by smnsht** [comment]
We have this situation covered in **Layout_FromMethod_FixedNullUrlValueTest()** (line 288)

```
[Fact]
public void Layout_FromMethod_FixedNullUrlValueTest()
{
    // Arrange
    Uri uri = null;
    Layout<Uri> layout = Layout<Uri>.FromMethod(l => uri);

    // Act
    var result = layout.RenderValue(LogEventInfo.CreateNullEvent());

    // Assert
    Assert.Equal(uri, result);
    Assert.Same(result, layout.RenderValue(LogEventInfo.CreateNullEvent()));
    Assert.Equal("", layout.Render(LogEventInfo.CreateNullEvent()));
    Assert.False(layout == uri);
    Assert.True(layout != uri);
    Assert.False(layout.IsFixed);
    Assert.NotEqual("null", layout.ToString());
    Assert.NotEqual(new Uri("//other"), layout);
}

```

**Comment by snakefoot** [comment]
The it seems covered already.

**Comment by smnsht** [comment]
Could it be AppVeyor build failing because of stripped-off trailing white spaces in the file headers?

```
Missing headers (copy them form other another file).
2 errors: 
 -------- 
-C:\projects\nlog\src/NLog\Layouts\Typed\Layout.cs
- C:\projects\nlog\tests/NLog.UnitTests\Layouts\LayoutTypedTests.cs 
Command exited with code 1
```

**Comment by snakefoot** [comment]
I think this initial API implementation looks good. Excellent job on the unit-tests to ensure the API behaves as expected.

Then future pull-request can work towards making it non-boxing.

**Comment by snakefoot** [comment]
Thank you for the contribution and getting the ball rolling. Created #5575 to continue the work.

**Pull request merged**